storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -889,60 +888,79 @@ public boolean pruneOldestBlobSidecars(
       return pruneBlobSidecars(
-          pruneLimit,
-          prunableNoncanonicalBlobKeys,
-          updater,
-          DatabaseArchiveNoopWriter.NOOP_BLOBSIDECAR_STORE,
-          true);
+          pruneLimit, prunableNoncanonicalBlobKeys, updater, archiveWriter, true);
     }
   }
 
   private boolean pruneBlobSidecars(
       final int pruneLimit,
       final Stream<SlotAndBlockRootAndBlobIndex> prunableBlobKeys,
       final FinalizedUpdater updater,
-      final DatabaseArchiveWriter<BlobSidecar> archiveWriter,
-      final boolean nonCanonicalblobSidecars) {
-    int remaining = pruneLimit;
+      final DataArchiveWriter<List<BlobSidecar>> archiveWriter,
+      final boolean nonCanonicalBlobSidecars) {
+
     int pruned = 0;
     Optional<UInt64> earliestBlobSidecarSlot = Optional.empty();
-    for (final Iterator<SlotAndBlockRootAndBlobIndex> it = prunableBlobKeys.iterator();
-        it.hasNext(); ) {
-      --remaining;
-      final boolean finished = remaining < 0;
-      final SlotAndBlockRootAndBlobIndex key = it.next();
-      // Before we finish we should check that there are no BlobSidecars left in the same slot
-      if (finished && key.getBlobIndex().equals(ZERO)) {
-        break;
+
+    // Group the BlobSidecars by slot. Potential for higher memory usage
+    // if it hasn't been pruned in a while
+    final Map<UInt64, List<SlotAndBlockRootAndBlobIndex>> prunableMap =
+        prunableBlobKeys.collect(groupingBy(SlotAndBlockRootAndBlobIndex::getSlot));
+
+    // pruneLimit is the number of slots to prune, not the number of BlobSidecars
+    final List<UInt64> slots = prunableMap.keySet().stream().sorted().limit(pruneLimit).toList();
+    for (final UInt64 slot : slots) {
+      final List<SlotAndBlockRootAndBlobIndex> keys = prunableMap.get(slot);
+
+      // Retrieve the BlobSidecars for archiving.
+      final List<BlobSidecar> blobSidecars =
+          keys.stream()
+              .map(
+                  nonCanonicalBlobSidecars
+                      ? this::getNonCanonicalBlobSidecar
+                      : this::getBlobSidecar)
+              .filter(Optional::isPresent)
+              .map(Optional::get)
+              .toList();
+
+      // Just warn if we failed to find all the BlobSidecars.
+      if (keys.size() != blobSidecars.size()) {
+        LOG.warn("Failed to retrieve BlobSidecars for keys:{}", keys);
       }
-      // Attempt to archive the BlobSidecar if present. If there's no BlobSidecar return true.
-      final boolean blobSidecarArchived =
-          getBlobSidecar(key).map(archiveWriter::archive).orElse(Boolean.TRUE);
+
+      // Attempt to archive the BlobSidecars.
+      final boolean blobSidecarArchived = archiveWriter.archive(blobSidecars);
       if (!blobSidecarArchived) {
-        LOG.warn("Failed to archive BlobSidecar for slot:{}. Stopping pruning", key.getSlot());
+        LOG.warn("Failed to archive and prune BlobSidecars. Stopping pruning");
```

Comment:
I wonder if this is an `error`. It requires an action from the user (like checking diskspace). We could also ask user to do something

Reply:
Yep. I agree. This would also mean LOG messages in FileSystemArchive should also be errors.